File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = []
        postfix = []
        pre_run = 1
        post_run = 1
        res = []
        for i in range(len(nums)):
            pre_run *= nums[i]
            prefix.append(pre_run)
        
        for i in range(len(nums)-1,-1,-1):
            post_run *= nums[i]
            postfix.append(post_run)
        
        postfix = postfix[::-1]
        
        print(prefix)
        print(postfix)

        while i < len(nums) - 1:
            if i-1 < 0:
                res.append(postfix[i+1])
            else:
                res.append(prefix[i-1]*postfix[i+1])
            i+=1
        res.append(prefix[len(nums)-2])
                
        return res



        

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len = nums.length;
        int[] left = new int[len];
        int[] right = new int[len];
        int[] res = new int[len];
        left[0] = nums[0];
        right[len-1] = nums[len-1];
        for(int i=1; i<len; i++) {
            left[i] = left[i-1]*nums[i];
        }
        for(int i=len-2; i>=0; i--) {
            right[i] = right[i+1]*nums[i];
        }

        for(int i=0; i<len; i++) {
            if(i == 0) {
                res[i] = right[i+1];
                continue;
            }
            if(i == len-1 ) {
                res[i] = left[i-1];
                continue;
            }
            res[i] = left[i-1]*right[i+1];
        }

        return res;

    }
}